// html-validate/dist/lexer/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenType = void 0;
var TokenType;
(function (TokenType) {
    TokenType[TokenType["UNICODE_BOM"] = 1] = "UNICODE_BOM";
    TokenType[TokenType["WHITESPACE"] = 2] = "WHITESPACE";
    TokenType[TokenType["NEWLINE"] = 3] = "NEWLINE";
    TokenType[TokenType["DOCTYPE_OPEN"] = 4] = "DOCTYPE_OPEN";
    TokenType[TokenType["DOCTYPE_VALUE"] = 5] = "DOCTYPE_VALUE";
    TokenType[TokenType["DOCTYPE_CLOSE"] = 6] = "DOCTYPE_CLOSE";
    TokenType[TokenType["TAG_OPEN"] = 7] = "TAG_OPEN";
    TokenType[TokenType["TAG_CLOSE"] = 8] = "TAG_CLOSE";
    TokenType[TokenType["ATTR_NAME"] = 9] = "ATTR_NAME";
    TokenType[TokenType["ATTR_VALUE"] = 10] = "ATTR_VALUE";
    TokenType[TokenType["TEXT"] = 11] = "TEXT";
    TokenType[TokenType["TEMPLATING"] = 12] = "TEMPLATING";
    TokenType[TokenType["SCRIPT"] = 13] = "SCRIPT";
    TokenType[TokenType["COMMENT"] = 14] = "COMMENT";
    TokenType[TokenType["CONDITIONAL"] = 15] = "CONDITIONAL";
    TokenType[TokenType["DIRECTIVE"] = 16] = "DIRECTIVE";
    TokenType[TokenType["EOF"] = 17] = "EOF";
})(TokenType = exports.TokenType || (exports.TokenType = {}));
